### PR TITLE
[GeoMechanicsApplication] Add thermal constitutive law

### LINKED ISSUE #11740: [GeoMechanicsApplication] Add thermal constitutive law
As a developer, I would like to add thermal constitutive law for my thermal elements such that I can build my conduction term.

### Description

**This feature is fully implemented.**

Branch: https://github.com/KratosMultiphysics/Kratos/tree/geo/10372-HeatTransfer-temp

- The conduction term of the heat equation is multiplied by dispersive matrix. It acts as a constitutive law for heat process. So, a constitutive law is added, GeoThermalDispersion2DLaw
applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp
applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Whole files

- The indices for the dispersive matrix are added (to be consistent with the other laws), INDEX_2D_THERMAL_FLUX_X, INDEX_2D_THERMAL_FLUX_Y and INDEX_2D_THERMAL_FLUX_Z
applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
lines 97 to 100

### Acceptance Criteria

**Given** the material parameter file with such a model which holds the data for this class
**When** checked for completeness and consistency once
**Then** gives the status

**Given** the material parameter file with such a model which holds the data for this class
**When** checked for completeness and consistency once
**Then** gives an exception error if the check fails

### PR DESCRIPTION
**Pull request by markelov208**
**📝 Description**
A brief description of the PR.

- added thermal constitutive law for the heat model

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- cleaning and refactoring thermal_dispersion_law.hpp/cpp files
- added unit tests for 2D and 3D cases

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: These local variables should have the lower_case format (see [style guide](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide#local-variables))

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: In general it is more clear if the result is returned instead of passed by reference and then changed by the function. In combination with a member variable that has the number of dimensions, we can just initialize it within the function and return it. 

This also prevents callers of this function from inputting an invalid matrix (like 4x2) and running into issues later on

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: Let's add your name here as well @markelov208

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp]: This comment is not correct, could you change it to 'Counted pointer of GeoThermalDispersion2DLaw' if you think this comment is informative (and otherwise remove it)?

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: It seems we should rename this (and the linked indexHeat2D) to remove the '2D' part, since it now deals with 2D and 3D

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: For the space_dimension, I would propose to have a member variable in the class, mNumberOfDimensions that we set in the constructor via arguments to either 2 or 3, such that this class knows its own dimensionality.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: There is some discussion with @mnabideltares @WPK4FEM @avdg81 about the naming of this class. Since this law is about the heat transfer, which at this point is only conduction, but will be extended with convection later on, I would say a more descriptive name would be e.g. GeoThermalTransferLaw, or GeoThermalConductionAndConvectionLaw if we really want to be specific. Let's discuss in office

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp]: Same here, let's add your name!

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: If we store the dimensionality in the class, we can just check that instead of this matrix.

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (rfaasse) [inline applications/GeoMechanicsApplication/geo_mechanics_application_constants.h]: Also here we can remove the '2D' as well as in the INDEX_2D_THERMAL_FLUX_X/Y/Z

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: Do you think it would be a good idea to have the thermal conductivity of solid and water always different, such that we also test there is no mix-up there?

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: Since it's a local variable, it should be lower_case
```suggestion
    Matrix thermal_dispersion_matrix = ZeroMatrix(2, 2);
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]:
```suggestion
    GeoThermalDispersion2DLaw geo_thermal_dispersion_2D_law;
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: Same here about the lower_case variables in line 72 and 73

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp]: For the SATURATION, @WPK4FEM proposed to investate if we could re-use the retention laws that are already there instead of having this new entry in the MaterialParameters.json. I think it would be a good idea to discuss this in office.

**Review by rfaasse — CHANGES_REQUESTED** [review]
Clear and understandable code and great to see the unit tests are added.

**Review by rfaasse — DISMISSED** [review]

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp]: This comment seems to be outdated.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp:1]: I think we should change the file extension to `.h`, since `.hpp` is reserved for files that provide both the declaration and the implementation.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: I believe we don't need this `#include` (perhaps it is a leftover from a debugging session?).

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: Shouldn't we need to add a check on `mNumberOfDimensions` here? Probably only the values 2 and 3 make sense here.
```suggestion
{
    KRATOS_ERROR_IF(mNumberOfDimensions < 2 || mNumberOfDimensions > 3) << "Got invalid number of dimensions: " << mNumberOfDimensions << std::endl;
}
```

File: applications/GeoMechanicsApplication/custom_retention/retention_law.h
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_retention/retention_law.h]: I would model a variable that may have an "unset" state using `std::optional`:
```suggestion
        std::optional<double> mFluidPressure;
```

File: applications/GeoMechanicsApplication/custom_retention/retention_law.h
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_retention/retention_law.h]: The prefix `r` is incorrect here, since it is not a reference.

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (avdg81) [inline applications/GeoMechanicsApplication/geo_mechanics_application_constants.h]: Please, use a scoped enum (i.e. `enum class`). See [Enum.3: prefer class enums over "plain" enums](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum3-prefer-class-enums-over-plain-enums) of the C++ Core Guidelines. This will also solve a code smell found by SonarCloud.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: I (Anne) shouldn't be mentioned as one of the main authors here. It was Gennady who did the hard work here :-)

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: The default value of the buffer size is 1, so we can leave it out here. (For this test case, that size doesn't matter, I think.)
```suggestion
    auto& r_model_part = current_model.CreateModelPart("ModelPart");
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: Since this is a test case for a 3D model, I would suggest to rename this variable:
```suggestion
    GeoThermalDispersionLaw geo_thermal_dispersion_3d_law(dimension);
```

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: I would also rename this local variable for consistency:
```suggestion
    GeoThermalDispersionLaw geo_thermal_dispersion_3D_law(dimension);
```
And perhaps also add a check (in this test case) for the 2D case.

**Review by avdg81 — COMMENTED** [review]
This PR looks nice and clean. I only have several minor comments and suggestions.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp]: removed @details. it seems it is a copy from another file.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.hpp:1]: renamed the file.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: good idea!

File: applications/GeoMechanicsApplication/custom_retention/retention_law.h
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_retention/retention_law.h]: fixed it.

File: applications/GeoMechanicsApplication/custom_retention/retention_law.h
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_retention/retention_law.h]: That is better.

File: applications/GeoMechanicsApplication/geo_mechanics_application_constants.h
Comment (markelov208) [reply applications/GeoMechanicsApplication/geo_mechanics_application_constants.h]: done, thank you for the info.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: removed. ;(

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: corrected

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: corrected

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: You are right, the line is removed.

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: 1 is removed

File: applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/tests/cpp_tests/test_thermal_dispersion_law.cpp]: the names are changed

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: I would replace the explicit type (`int`) by `auto`, since it is repeating the type expressed by the `static_cast<int>`. SonarQube will report code smells for these as well.
```suggestion
    const auto x = static_cast<int>(indexThermalFlux::X);
    const auto y = static_cast<int>(indexThermalFlux::Y);
```

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: And the same thing here:
```suggestion
        const auto z = static_cast<int>(indexThermalFlux::Z);
```

**Review by avdg81 — DISMISSED** [review]
This looks really nice. Thanks for the hard work Gennady!

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: changed.

File: applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_law.cpp]: thank you for the info. changed.

**Comment by markelov208** [comment]
Hi Anne, thank you for the info on code smells and high esteem of my work.

**Review by avdg81 — APPROVED** [review]
I think we're ready to merge. Please Gennady, feel free to press the button :-)

**Comment by markelov208** [comment]
Good morning Anne, unfortunately there is a failed test, MeshMovingApplication. I think this is not GeoMechanics but it blocks the merge button. Is there a 'secret' button to accept failed tests? GitLab has it.

**Comment by avdg81** [comment]
> Good morning Anne, unfortunately there is a failed test, MeshMovingApplication. I think this is not GeoMechanics but it blocks the merge button. Is there a 'secret' button to accept failed tests? GitLab has it.

Yes, we've noticed ;-) I have pressed the button to rerun the failed jobs. Not sure whether that will help, but let's try. I'm not aware of any "secret" button to accept failed tests.

**Pull request merged**
